jupyterlab_chameleon/artifact.py

```diff
@@ -1,66 +1,257 @@
+    """A shared experiment/research artifact that can be spawned in JupyterHub.
+
+    Attrs:
+        deposition_repo (str): the name of the deposition repository (e.g.,
+            "zenodo" or "chameleon")
+        deposition_id (str): the ID of the deposition within the repository.
+        id (str): the external Trovi ID of the artifact linked to this
+            deposition. Default = None.
+        tags (list[str]): a list of tags used to categorize this artifact.
+        authors (list[ArtifactAuthor]): a list of the authors of this artifact
+        linked_projects (list[str]): a list of URNs representing projects to which
+            this artifact is linked.
+        repro_enable_requests (bool): flag which allows external users to request
+            to reproduce this artifact's experiment. Default = False
+        repro_access_hours (int): the number of hours given to external users to
+            reproduce this artifact's experiments.
+        repro_requests (int): the number of reproduction requests made to this artifact
+        versions (list[ArtifactVersion]): a list of all versions of this artifact
+        ownership (str): the requesting user's ownership status of this
+            artifact. Default = "fork".
+    """
+
+    def __init__(
+        self,
+        id=None,
+        tags=None,
+        authors=None,
+        linked_projects=None,
+        repro_enable_requests=False,
+        repro_access_hours=None,
+        repro_requests=None,
+        versions=None,
+    ):
+        self.id = id
+        self.tags = tags or []
+        self.authors = authors or []
+        self.linked_projects = linked_projects or []
+        self.repro_enable_requests = repro_enable_requests
+        self.repro_access_hours = repro_access_hours
+        self.repro_requests = repro_requests
+        self.versions = versions or []
+
+        # Only the deposition information is required. Theoretically this can
+        # allow importing arbitrary Zenodo DOIs or from other sources that are
+        # not yet existing on Trovi.
+        if not versions:
+            raise ValueError("Missing artifact contents")
+
+
 def default_prepare_upload():
     """Prepare an upload to the external storage tier.
 
     Returns:
         dict: a structure with the following keys:
-            :deposition_id: the pre-prepared ID of the new artifact deposition
-            :publish_endpoint: the upload endpoint parameters:
+            :upload_endpoint: the upload endpoint parameters:
                 :url: the absolute URL to perform the upload
                 :method: the request method
                 :headers: any headers to include in the request
     """
-    return call_jupyterhub_api('share/prepare_upload')
+    trovi_token = get_trovi_token()
+
+    response = {
+        "upload_endpoint": {
+            "url": contents_url(trovi_token),
+            "method": "POST",
+            "headers": {
+                "content-type": "application/json",
```

Comment:
content-type is not JSON for an upload, won't it be a binary upload of a tarball?